### DISCUSSION TITLE
Adopt the .NET8 IPNetwork instead of .NET6+ IPAddressRange

### DISCUSSION THREAD

**Discussion by WeihanLi** [ThreeMammals/Ocelot#1799] [Ideas]
.NET 8 introduced the `IPNetwork`, maybe we could use `IPNetwork` instead for `net8.0`

- [IPNetwork Struct](https://learn.microsoft.com/en-us/dotnet/api/system.net.ipnetwork?view=net-8.0)

**Comment by raman-m** [comment]
Hello, Weihan!
Welcome to Ocelot world from YARP one! 🤣 

And how to support `net7.0` and `net6.0` ?
As you said, this new feature is available for `net8.0` only.

If you find out this interesting, who will develop? 🥲

**Reply by ks1990cn** [reply]
@raman-m before this we have existing which is doing its job till .NET 8, we would be using if we go to .Net 8, I have question on it, do @WeihanLi wants something we haven't imagined?

**Reply by WeihanLi** [reply]
Yep, I meant for .NET 8 we could use the new `IPNetwork`, for `net6.0`/`net7.0` we could keep using `IPAddressRange`, so that we could keep our library clean and less dependencies

Greatly happy that Ocelot continues and get more active, thank you very much to you guys for making it happen

**Reply by raman-m** [reply]
You are welcomed to Ocelot world again! I know you are [old contributor](../commits?author=WeihanLi) of our repo. :heart: 

We've prepared new release after v.21:
- #1805 

Regarding `IPNetwork`... What's the benefit here? :ok:  We've migrated to use `IPNetwork`, but what is a benefit for Ocelot?

**Comment by ks1990cn** [comment]
Any specific reason for this suggestion? I can see IPAddressRange is supported till .Net 8.

https://www.nuget.org/packages/IPAddressRange/#supportedframeworks-body-tab

**Comment by ggnaegi** [comment]
Hello

... too late, the others replied first...

I didn't know that Microsoft was offering this feature in .NET 8, but this implementation isn't supported in .NET 7 and .NET 6, whereas IPAdressRange is... 

There are lots of interesting new features introduced with .NET 7 and .NET 8, such as the new rate limiters - I'll have to check whether it's now possible to combine them -. I think @RaynaldM would also like to see the new rate limiters integrated into the solution :-)

First of all we need to define the strategy for the future in order to avoid an out-of-control architecture (preprocessors all over the place)

**Reply by raman-m** [reply]
Aha, Rate Limiting redesigning has high priority! 😉 since ASP.NET 7 and Ocelot 19.0.3 😄

**Reply by raman-m** [reply]
There is no problem to extract new feature interfaces and to ILDasm the concreate classes, but I have no time for such design adventures. 🤣 
But @WeihanLi has the time 🔥

**Reply by ggnaegi** [reply]
Maybe with a bit of preprocessors and DI...

**Reply by raman-m** [reply]
To be fair, @WeihanLi's idea is pretty abstract...
He needs to develop his idea more and show us real code snippets.

**Reply by RaynaldM** [reply]
To introduce new Net8 features (such as rate limiting), you can use the DI with directives such as :
```
    public static IOcelotBuilder AddPolly(this IOcelotBuilder builder)
    {
#if NET8_0_OR_GREATER
return builder;
#elif NET7_0
return builder;
#else // Net6.0
        var errorMapping = new Dictionary<Type, Func<Exception, Error>>
        {
            { typeof(TaskCanceledException), e => new RequestTimedOutError(e) },
            { typeof(TimeoutRejectedException), e => new RequestTimedOutError(e) },
            { typeof(BrokenCircuitException), e => new RequestTimedOutError(e) },
            { typeof(BrokenCircuitException<HttpResponseMessage>), e => new RequestTimedOutError(e) },
        };
        return AddPolly<PollyQoSProvider>(builder, GetDelegatingHandler, errorMapping);
#endif
    }
```
In the code itself, you can also use this kind of directive, but be careful not to overuse it, as it can be a pain to maintain and test.
![image](https://github.com/ThreeMammals/Ocelot/assets/606401/0467bd31-f685-4920-821f-b7bff5f080ef)

To switch from one .Net version to another, click here
![image](https://github.com/ThreeMammals/Ocelot/assets/606401/6fe61fcd-04b0-4115-a347-37541b77627f)

Normally unit tests are automatically passed for each .Net version.

**Reply by raman-m** [reply]
Wow! Ray is online! 💥 Lunch time? 😉

**Reply by ggnaegi** [reply]
@raman-m @RaynaldM for features like the Rate limiters, I would go a step further for clarity, such as
```c#
#if NET7_0_OR_GREATER
            Services.TryAddSingleton<IRequestMapper, RequestMapper>();
#else
            Services.TryAddSingleton<IRequestMapper, RequestMapper>();
#endif
```
Trying to avoid the preprocessors in the interface implementations.

**Reply by raman-m** [reply]
Preprocessor directives won't be required if we develop native feature and encapsulate `.net8.0` interfaces...
But I'm afraid, having .NET 6-7-8 multitargeting requires to have some preprocessing by compiler. Well... This is design challenge.

**Comment by RaynaldM** [comment]
" I think @RaynaldM would also like to see the new rate limiters integrated into the solution :-)"

Not only that 😃

**Reply by raman-m** [reply]
Ray, we are not your personal development team for free! :rofl:  :wink:
